core/src/KokkosExp_InterOp.hpp

```diff
@@ -0,0 +1,172 @@
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_CORE_EXP_INTEROP_HPP
+#define KOKKOS_CORE_EXP_INTEROP_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+#include <Kokkos_Layout.hpp>
+#include <Kokkos_MemoryTraits.hpp>
+#include <Kokkos_View.hpp>
+#include <Kokkos_DynRankView.hpp>
+#include <impl/Kokkos_Utilities.hpp>
+#include <type_traits>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+template <typename... T>
+using type_list = Kokkos::Impl::type_list<T...>;
+
+// ------------------------------------------------------------------ //
+//  concat_type_list combines types in multiple type_lists
+
+// default definition
+template <typename... T>
+struct concat_type_list {
+  using type = type_list<T...>;
+};
+
+// append to type_list
+template <typename... T, typename... Tail>
+struct concat_type_list<type_list<T...>, Tail...>
+    : concat_type_list<T..., Tail...> {};
+
+// combine consecutive type_lists
+template <typename... T, typename... U, typename... Tail>
+struct concat_type_list<type_list<T...>, type_list<U...>, Tail...>
+    : concat_type_list<type_list<T..., U...>, Tail...> {};
+
+// alias
+template <typename... T>
+using concat_type_list_t = typename concat_type_list<T...>::type;
+
+// ------------------------------------------------------------------ //
+//  gather filters out types which satisfy PredicateT<T>::Value = ValueT
+
+template <template <typename> class PredicateT, bool ValueT, typename... T>
+struct gather {
+  using type = concat_type_list_t<std::conditional_t<
+      PredicateT<T>::value == ValueT, concat_type_list_t<T>, type_list<>>...>;
+};
```

Comment:
I find it weird that you have both a predicate and a value.
Also I would prefer `template <template <class> class Predicate, class TypeList> filter_type_list`

Reply:
I've found in the past that it is usually worth having a predicate and value in the long run bc what if you want to gather the types where the predicate is true?

Reply:
If we put the type list first in the template parameters, we can then have the value as an optional third parameter, as in `template <class Typelist, template <class> class Predicate, bool PredValue = true> struct filter_type_list`.

Reply:
> If we put the type list first in the template parameters, we can then have the value as an optional third parameter, as in template <class Typelist, template <class> class Predicate, bool PredValue = true> struct filter_type_list.

@nliber Are you sure about this? I am expanding the `std::conditional_t<...>` for `T...`, which `class Typelist` lacks.

Reply:
`ValueT` is now default parameter:

```cpp
template <template <typename> class PredicateT, typename T, bool ValueT = false>
using gather_t = typename gather<PredicateT, T, ValueT>::type;
```